design/baremetal-operator/bmh-v1beta1.md

```diff
@@ -0,0 +1,126 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# BareMetalHost API v1beta1
+
+## Status
+
+One of: implementable
+
+## Summary
+
+This design proposes making certain backward incompatible changes to the
+BareMetalHost API and publishing the result as version `v1beta1`.
+
+## Motivation
+
+The BareMetalHost API is currently stuck at `v1alpha1`. Meanwhile, several
+incompatible changes are needed based on the recent development both in Metal3
+itself and in Ironic.
+
+For example, Ironic is planning a complete removal of MD5 support for
+checksums.
+
+### Goals
+
+- BareMetalHost API version `metal3.io/v1beta1`.
+
+### Non-Goals
+
+- Stabilize the BareMetalHost API.
+
+## Proposal
+
+Copy the current API types definition into a new package `v1beta1` with some
+modifications.
+
+## Design Details
+
+The following addition should be done *before* the version is created:
+
+- Add a `Spec.Architecture` field to override architecture in the profile.
+  If not provided, it will be populated after inspection.
+
+The following changes will be done to the new version:
+
+- Remove MD5 support from `Image.ChecksumType`. When no `ChecksumType` is
+  provided, do not populate the `os_image_hash_algo` field and let Ironic
+  guess the algorithm.
+
+  The same change is applied to CAPM3.
+
+- In `RootDeviceHints`:
+
+  Rename `DeviceName` to `DevicePath` to better reflect what it actually is
+  (especially after introducing support for `/dev/disk/by-path`).
+
+  Rename `Model` and `Vendor` to `ModelContains` and `VendorContains` to
+  reflect the fact that these fields use the `<in>` operator.
+
+- Remove `HardwareProfile` and don't do any matching.
+
+- Remove `HardwareDetails` from the `Status` in favour of the already
+  introduced `HardwareData` CR.
+
+### Implementation Details/Notes/Constraints
+
+Both `v1alpha1` and the proposed `v1beta1` resources will be possible
+to handle with the same controller code since the differences are mostly
+cosmetic.
+
+The new architecture handling will make it possible for a
+`PreprovisioningImage` to have an empty architecture requested (currently,
+it defaults to `x86_64`). We expect implementations to use its default
+in this case. The `BareMetalHost` controller will start handling the situation
+where requested architecture is empty and provided architecture is not.
+
+### Risks and Mitigations
+
+None
+
+### Work Items
+
+- Add `Architecture`
+- Copy type definitions and modify them.
```

Comment:
We also have to find the 5000 places that `v1alpha1` appears in the controller code and change them. Preferably to something that includes the version in far fewer places (like only in the import lines).

Reply:
Yes, +1 to avoid the version. This is what CAPI and many of the providers do also. In CAPM3 we rename the import `infrav1` ([example](https://github.com/metal3-io/cluster-api-provider-metal3/blob/6a9fe1d9e9fbea791094a709d5e79ad99c63b34b/controllers/metal3cluster_controller.go#L27)). Something similar should work well also for BMO. :slightly_smiling_face: